File: android/app/src/main/java/com/vergipro/mobile/feature/capture/CaptureSubmissionService.kt
package com.vergipro.mobile.feature.capture

import com.vergipro.mobile.core.sync.OfflineMutationQueue
import com.vergipro.mobile.core.sync.OfflineOperation
import com.vergipro.mobile.core.sync.QueuedMutation
import java.io.ByteArrayOutputStream
import java.io.DataOutputStream
import java.time.Instant
import java.util.UUID

class CaptureSubmissionService(private val queue: OfflineMutationQueue) {
    fun prepare(pages: List<CapturedPage>, organizationId: Int, submissionId: UUID): UUID {
        require(pages.isNotEmpty() && pages.all { it.vaultDocument.organizationId == organizationId })
        val payload = ByteArrayOutputStream().use { bytes ->
            DataOutputStream(bytes).use { output ->
                output.writeInt(1)
                output.writeUTF(submissionId.toString())
                output.writeInt(organizationId)
                output.writeLong(Instant.now().toEpochMilli())
                output.writeInt(pages.size)
                pages.forEach { output.writeUTF(it.vaultDocument.id.toString()) }
            }
            bytes.toByteArray()
        }
        queue.enqueue(QueuedMutation(submissionId, organizationId, submissionId.toString(), OfflineOperation.UploadDocument, payload, submissionId))
        return submissionId
    }

    fun status(submissionId: UUID): QueuedMutation? = queue.status(submissionId)
}

File: ios/VergiPro/Features/Capture/CaptureSubmissionService.swift
import Foundation

struct CaptureSubmissionPayload: Codable, Sendable {
    let submissionID: UUID
    let organizationID: Int
    let createdAt: Date
    let pages: [VaultDocument]
}

actor CaptureSubmissionService {
    private let queue: OfflineMutationQueue

    init(queue: OfflineMutationQueue) { self.queue = queue }

    func prepare(pages: [CapturedPage], organizationID: Int, submissionID: UUID) async throws -> UUID {
        guard !pages.isEmpty, pages.allSatisfy({ $0.vaultDocument.organizationID == organizationID }) else {
            throw APIError.invalidRequest
        }
        let payload = CaptureSubmissionPayload(submissionID: submissionID, organizationID: organizationID, createdAt: .now, pages: pages.map(\.vaultDocument))
        let mutation = QueuedMutation(
            id: submissionID, organizationID: organizationID, aggregateID: submissionID.uuidString,
            operation: .uploadDocument, payload: try JSONEncoder().encode(payload), idempotencyKey: submissionID,
            attemptCount: 0, nextAttemptAt: nil, state: .pending
        )
        try await queue.enqueue(mutation)
        return submissionID
    }

    func status(submissionID: UUID) async -> QueuedMutation? { await queue.status(id: submissionID) }
}
